### PR TITLE
Editor state management

### PR DESCRIPTION
**Pull request by mrniket**
This change is [Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/555)

### CODE REVIEW

**Comment by CelineBoudier** [comment]
Reviewed 5 of 32 files at r1, 1 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[game_frontend/src/components/GamePage/index.test.js, line 4 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQAxAzDBr4KJ90RSQm:-LBQEbrtCm77qRkj0CIY:b32v57) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/components/GamePage/index.test.js#L4)):*
<details><summary><i>Previously, OlafSzmidt (Olaf Szmidt) wrote…</i></summary><blockquote>

> ```
> // import toJson from 'enzyme-to-json'
> ```

Commented import
</blockquote></details>

+1

---


*Comments from [Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/555)*

**Comment by CelineBoudier** [comment]
Reviewed 15 of 32 files at r1, 4 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[game_frontend/src/redux/features/Editor/types.js, line 10 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQHq389-anjWiFPhcY:-LBQHq389-anjWiFPhcZ:bsnhvy9) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/redux/features/Editor/types.js#L10)):*
> ```JavaScript
> 
> const CHANGE_CODE = 'features/Editor/CHANGE_CODE'
> const EDITOR_CHANGED = 'features/Editor/EDITOR_CHANGED'
> ```

what does editor_changed mean? Trying to understand the naming pattern compared to change_code

---


*Comments from [Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/555)*

**Comment by OlafSzmidt** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[game_frontend/src/redux/features/Editor/types.js, line 10 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQHq389-anjWiFPhcY:-LBRLM7_1h9Nr33gz-rB:b-zhzsp) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/redux/features/Editor/types.js#L10)):*
<details><summary><i>Previously, CelineBoudier (Celine Boudier) wrote…</i></summary><blockquote>

what does editor_changed mean? Trying to understand the naming pattern compared to change_code
</blockquote></details>

```what does editor_changed mean? Trying to understand the naming pattern compared to change_code?```

Basically it's a type used by an action. Every time you press something in the editor, we send a `CHANGE_CODE` action, and the epic will debounce only ever so often (300ms), and send a `EDITOR_CHANGED` action to the state. Saves us changing the state multiple times a second when typing.

That being said, the first word should be a verb :+1: 
```


---


*Comments from [Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/555)*
<!-- Sent from Reviewable.io -->

**Comment by mrniket** [comment]
Review status: 26 of 27 files reviewed at latest revision, 5 unresolved discussions.

---

*[game_frontend/src/containers/IDEMenu/index.js, line 28 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQCoP83A9lIGdfjkVB:-LBV5-LH383X6ESHSH_-:b-ze1p34) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/containers/IDEMenu/index.js#L28)):*
<details><summary><i>Previously, OlafSzmidt (Olaf Szmidt) wrote…</i></summary><blockquote>

> ```
> () => ({})
> ```
Following the pattern in `mapDispatchToProps`, maybe we can try making this an empty object too? Shorthand notation: https://egghead.io/lessons/javascript-redux-using-mapdispatchtoprops-shorthand-notation
</blockquote></details>

It didn't work for mapStateToProps :(

---

*[game_frontend/src/redux/features/Editor/types.js, line 10 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQHq389-anjWiFPhcY:-LBV5Rv9BNZ-D-ZOqZKM:ba5gedz) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/redux/features/Editor/types.js#L10)):*
<details><summary><i>Previously, OlafSzmidt (Olaf Szmidt) wrote…</i></summary><blockquote>

```what does editor_changed mean? Trying to understand the naming pattern compared to change_code?```

Basically it's a type used by an action. Every time you press something in the editor, we send a `CHANGE_CODE` action, and the epic will debounce only ever so often (300ms), and send a `EDITOR_CHANGED` action to the state. Saves us changing the state multiple times a second when typing.

That being said, the first word should be a verb :+1: 
</blockquote></details>

actually Olaf it's the other way around, the `EDITOR_CHANGED` is sent by the editor and then the `CHANGE_CODE` action is what results in the `state.editor.code` updating. The reason the verb is at the end of the action `EDITOR_CHANGED` is that the action happens more of a report of what's happened vs the other actions where there is an intent for something to happen as a result of the action. However, if you guys think `CHANGED_EDITOR` makes more sense I'm happy to change it :)

---


*Comments from [Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/555)*

**Comment by mrniket** [comment]
Review status: 26 of 27 files reviewed at latest revision, 5 unresolved discussions.

---

*[game_frontend/src/containers/IDEEditor/index.js, line 52 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQBWKb-TS3Ms-iNr-O:-LBV6iqG6j5E_atZ4LIC:b-896fix) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/containers/IDEEditor/index.js#L52)):*
<details><summary><i>Previously, OlafSzmidt (Olaf Szmidt) wrote…</i></summary><blockquote>

Let's use shorthand to avoid boilerplate like below
</blockquote></details>

Done.

---

*[game_frontend/src/redux/features/Editor/epics.test.js, line 122 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQDQoA9WWrrOj6JvB2:-LBV6kxj1llmAp6NOjqO:bf4w8ri) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/redux/features/Editor/epics.test.js#L122)):*
<details><summary><i>Previously, OlafSzmidt (Olaf Szmidt) wrote…</i></summary><blockquote>

> ```
> a certain amount
> ```
More accurate description
</blockquote></details>

is this better?

---


*Comments from [Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/555)*

**Comment by OlafSzmidt** [comment]
Reviewed 3 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[game_frontend/src/redux/features/Editor/epics.test.js, line 122 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQDQoA9WWrrOj6JvB2:-LBVCGi97dOL2kDBz2N3:b-89q77x) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/redux/features/Editor/epics.test.js#L122)):*
<details><summary><i>Previously, mrniket (Niket Shah) wrote…</i></summary><blockquote>

is this better?
</blockquote></details>

Cool!

---

*[game_frontend/src/redux/features/Editor/types.js, line 10 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQHq389-anjWiFPhcY:-LBVCBWa2mFybQ2Yrnhs:b9y1bfs) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/redux/features/Editor/types.js#L10)):*
<details><summary><i>Previously, mrniket (Niket Shah) wrote…</i></summary><blockquote>

actually Olaf it's the other way around, the `EDITOR_CHANGED` is sent by the editor and then the `CHANGE_CODE` action is what results in the `state.editor.code` updating. The reason the verb is at the end of the action `EDITOR_CHANGED` is that the action happens more of a report of what's happened vs the other actions where there is an intent for something to happen as a result of the action. However, if you guys think `CHANGED_EDITOR` makes more sense I'm happy to change it :)
</blockquote></details>

Yes please

---


*Comments from [Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/555)*

**Comment by CelineBoudier** [comment]
Reviewed 6 of 32 files at r1, 3 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[game_frontend/src/redux/features/Editor/types.js, line 10 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQHq389-anjWiFPhcY:-LBVU_FAEoVDdq0YxT6y:b-fr1eec) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/redux/features/Editor/types.js#L10)):*
<details><summary><i>Previously, OlafSzmidt (Olaf Szmidt) wrote…</i></summary><blockquote>

Yes please
</blockquote></details>

Niket's answer makes sense. I'm happy either way tbh

---


*Comments from [Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/555)*

**Comment by mrniket** [comment]
Review status: 22 of 27 files reviewed at latest revision, 1 unresolved discussion.

---

*[game_frontend/src/redux/features/Editor/types.js, line 10 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/555#-LBQHq389-anjWiFPhcY:-LBVZ5B90noz4h-1H1nr:bmklx76) ([raw file](https://github.com/ocadotechnology/aimmo/blob/2a2b4fea31ee1a4605bbca6e4ff655f8017b8ebd/game_frontend/src/redux/features/Editor/types.js#L10)):*
<details><summary><i>Previously, CelineBoudier (Celine Boudier) wrote…</i></summary><blockquote>

Niket's answer makes sense. I'm happy either way tbh
</blockquote></details>

changed it to KEY_PRESSED as discussed, makes more sense I think now! 

---


*Comments from [Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/555)*
